enhancements/api-review/add-mirror-by-digest-only.md

```diff
@@ -0,0 +1,226 @@
+approvers:
+  - TBD
+creation-date: 2021-03-10
+last-updated: 2021-03-10
+status: implementable
+---
+
+# Add MirrorByDigestOnly to cluster wide ImageContentSourcePolicy
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today, the ImageContentSourcePolicy object sets up mirror configuration with `mirror-by-digest-only` property set to true. This enhancement plans to make `mirror-by-digest-only` a configurable option.
+
+## Motivation
+
+Today, the ImageContentSourcePolicy object sets up mirror configuration with `mirror-by-digest-only` property set to true, which leads to using mirrors only when images are referenced by digests. However, when working with disconnected environments, sometimes multiple ImageContentSourcePolicies are needed, some of them are used by apps/manifests that don't use digests when pulling the images.
+
+Adding the `mirror-by-digest-only` option to ImageContentSourcePolicy will make it possible to pull images from mirror using tags without requiring image digest reference.
+
+### Goals
+
+- Enable the user to be able to set `mirror-by-digest-only` through ImageContentSourcePolicy object.
+
+### Non-Goals
+
+## Proposal
+
+The Image API has an optional `MirrorByDigestOnly` filed with type `bool` to `Registry`.
+
+```go
+// Registry represents a registry.
+type Registry struct {
+  // ...
+
+	// If true, mirrors will only be used for digest pulls. Pulling images by
+	// tag can potentially yield different images, depending on which endpoint
+	// we pull from.  Forcing digest-pulls for mirrors avoids that issue.
+	MirrorByDigestOnly bool `toml:"mirror-by-digest-only,omitempty"`
+}
+```
+
+The ImageContentSourcePolicy CRD has the cluster-wide information about handle the mirrors only when pulling the images that are referenced by their digests.
+It will now watch for `mirror-by-digest-only` to not require digest reference.
+
+An example ImageContentSourcePolicy file will look like:
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: ImageContentSourcePolicy
+metadata:
+  name: ubi8repo
+spec:
+  repositoryDigestMirrors:
+  - mirrors:
+    - example.io/example/ubi-minimal 
+    source: registry.access.redhat.com/ubi8/ubi-minimal
+    mirror-by-digest-only: false
+```
+
+### User Stories
+
+#### As a user, I would like to pull images from mirror using tag, without digests reference
+The user can specify `mirror-by-digest-only: false` for the `source` under `repositoryDigestMirrors` in the ImageContentSourcePolicy file. And create the ImageContentSourcePolicy project. Once this is done, the images can be pulled from the mirrors without the digest referenced.
```

Comment:
This use-case should probably be expanded to explain why anybody would want to do this.  Unless you also have a way to mirror signatures (which sadly are still not something that gets represented in the image's blog graph, as far as I'm aware), it exposes you to man-in-the-middle attacks from anywhere between the image-build location and your image-consuming cluster.  For example, the image is vulnerable to replacement with a malicious image while transitioning between the build system and the mirror repository, while it is at rest on the mirror repository, and while it is transitioning between the mirror repository and the node about to launch a container.  Some of those transitions might be covered by TLS, but still, you need a lot of trust in all the steps in the chain of custody to be able to do this with any confidence.  Is transitioning consumers to by-digest pullspecs really that difficult?